proposals/trampoline.md

```diff
@@ -0,0 +1,346 @@
+
+```text
+                        update_add_htlc
++---------------------------------------------------------------------------+
+| channel_id                                                                |
++---------------------------------------------------------------------------+
+| htlc_id                                                                   |
++---------------------------------------------------------------------------+
+| amount_msat                                                               |
++---------------------------------------------------------------------------+
+| payment_hash                                                              |
++---------------------------------------------------------------------------+
+| cltv_expiry                                                               |
++---------------------------------------------------------------------------+
+|               onion_routing_packet                                        |
+| +-----------------------------------------------------------------------+ |
+| | version                                                               | |
+| +-----------------------------------------------------------------------+ |
+| | public_key                                                            | |
+| +-----------------------------------------------------------------------+ |
+| | +--------------------------------------------+----------------------+ | |
+| | | amt_to_forward                             |                      | | |
+| | +--------------------------------------------+                      | | |
+| | | outgoing_cltv_value                        | normal onion payload | | |
+| | +--------------------------------------------+ for Irvin            | | |
+| | | short_channel_id                           |                      | | |
+| | +--------------------------------------------+----------------------+ | |
+| | +--------------------------------------------+----------------------+ | |
+| | | amt_to_forward                             |                      | | |
+| | +--------------------------------------------+                      | | |
+| | | outgoing_cltv_value                        | normal onion payload | | |
+| | +--------------------------------------------+ for Iris             | | |
+| | | short_channel_id                           |                      | | |
+| | +--------------------------------------------+----------------------+ | |
+| | +--------------------------------------------+----------------------+ | |
+| | | amt_to_forward                             |                      | | |
+| | +--------------------------------------------+                      | | |
+| | | outgoing_cltv_value                        |                      | | |
+| | +--------------------------------------------+                      | | |
+| | | payment_secret                             |                      | | |
```

Comment:
I'm not sure I understand why trampoline nodes should require access to the `payment_secret`, shouldn't that be in the trampoline onion, destined for the recipient, or at most the last trampoline before the destination that doesn't support trampoline only? If the latter was the intended semantics it'd need a note to this regard.

Reply:
Or are they intended to be decoy secrets?

Reply:
I should make this clearer, there are two distinct `payment_secret`s here to allow MPP to be aggregated at trampoline nodes before being re-split by the trampoline node when forwarding.

Let me know if these ASCII arts are better: https://github.com/ACINQ/eclair/blob/ce3629c98b07cdc914bd397fe45dd4bfd8338392/eclair-core/src/main/scala/fr/acinq/eclair/wire/Onion.scala#L59

The idea is that we need to support MPP payments split between multiple different trampoline routes (those need a `payment_secret` in the inner, trampoline onion) and using MPP between trampoline nodes to forward trampoline payments efficiently (those needs a different, unrelated `payment_secret` in the outer, normal onion).

Reply:
Ah ok, that makes sense, but wouldn't that also be possible just by aggregating across the `payment_hash`? Barring accidental reuse of the hash for different destinations (which is inherently unsafe anyway) it should work just as well.

FWIW even the concurrent reuse case is just a special case of the general reuse case, so I don't think it needs special treatment.

Reply:
That would work but you'd lose the probing protection that the `payment_secret` brings, so a node Mallory that's just before a trampoline node Ted could guess that Ted is either the payment final recipient or a trampoline node. Not sure they could steal fees, but they might want to try which is bad for the payment success rate, so it felt better to keep the `payment_secret` mechanism.